saleor/product/utils/attributes.py

```diff
@@ -81,3 +81,12 @@ def associate_attribute_values_to_instance(
     assignment = _associate_attribute_to_instance(instance, attribute.pk)
     assignment.values.set(values)
     return assignment
+
+
+def unpublished_products_before_category_delete(category):
+    products = category.products.all()
+    descendants = category.get_descendants()
+    for descendant in descendants:
+        products = products | descendant.products.all()
+    products.update(is_published=False, publication_date=None)
+    return products
```

Comment:
That's not a utility for attributes (cf. the filename).

Also, you could add it as a manager method on the category model, would be cleaner 😉

Reply:
It'd fine to just move it to `saleor/product/__init__.py` which contains some general utility functions related to products.

Reply:
Better to avoid `__init__` it is not supposed to contain >10% of the logic, we expect mostly imports in it.

Reply:
But it does at the moment. It's better to put the function there now and do the refactoring in a separate PR, since we would need to probably move more functions.